src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -737,5 +737,62 @@ public static TimeSpan GetTotalPauseDuration()
         {
             return new TimeSpan(_GetTotalPauseDuration());
         }
+
+        [ThreadStatic]
```

Comment:
Instead of thread static, you should use `void* context` argument that gets passed as an argument to the method and back to the callback. It is the standard pattern to communicate with callbacks in C/C++.

Reply:
@jkotas Are you thinking of a GCHandle to the dictionary or a pointer to a ref struct that holds the dictionary?

Reply:
We typically use pointer to a struct for synchronous callbacks. Here are a few example: https://github.com/dotnet/runtime/blob/4017327955f1d8ddc43980eb1848c52fbb131dfc/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Icu.cs#L430
https://github.com/dotnet/runtime/blob/d6e8686e566efbf9e921581abbfb2dc114369168/src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/UnixIPGlobalProperties.cs#L116

We typically use regular struct for the context even though the struct is really a ref struct. If you turn the struct into `ref struct`, it is not possible to use `Unsafe.AsPointer` anymore and the whole thing gets complicated.